### PR TITLE
Feature/note lane distribution

### PR DESCRIPTION
**Pull request by PiterGroot**
https://github.com/lucasvdiepen/ProefExamen/assets/70685433/99ae3249-f9c8-49e4-9da4-fb0d60b7a5f0

### CODE REVIEW

File: Assets/Scripts/Framework/BeatMapping/AudioWaveformDrawer.cs [hunk 1]
```diff
@@ -1,5 +1,7 @@
 using UnityEngine;
 using System;
+using ProefExamen.Framework.Gameplay.LaneSystem;
+using ProefExamen.Framework.Gameplay.Values;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/AudioWaveformDrawer.cs]:
```suggestion
using UnityEngine;
using System;

using ProefExamen.Framework.Gameplay.LaneSystem;
using ProefExamen.Framework.Gameplay.Values;
```

File: Assets/Scripts/Framework/BeatMapping/TimeStampData.cs
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampData.cs:1]: Put class into correct namespace.

File: Assets/Scripts/Framework/BeatMapping/TimeStampDataContainer.cs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+using ProefExamen.Framework.Gameplay.Level;
 using System.Collections.Generic;
 using UnityEngine;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampDataContainer.cs]:
```suggestion
using System.Collections.Generic;
using UnityEngine;

using ProefExamen.Framework.Gameplay.Level;
```

File: Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 #if UNITY_EDITOR
+
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs]:
```suggestion
```

File: Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs [hunk 4]
```diff
@@ -20,13 +22,38 @@ public class TimeStampDebugger : MonoBehaviour
 
         [SerializeField]
         private float _gizmoSpacing = .6f;
+        
+        [SerializeField]
+        private Toggle[] laneToggles;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs]:
```suggestion
        private Toggle[] _laneToggles;
```

File: Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs [hunk 5]
```diff
@@ -20,13 +22,38 @@ public class TimeStampDebugger : MonoBehaviour
...
         private AudioWaveformDrawer _waveformDrawer;
 
         private readonly GUIStyle _debugBoldGuiStyle = new();
         private readonly GUIStyle _debugItalicsGuiStyle = new();
 
+        private bool lane1Enabled = true;
+        private bool lane2Enabled = true;
+
+        private bool lane3Enabled = true;
+        private bool lane4Enabled = true;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs]:
```suggestion
        private bool _firstLaneEnabled = true;
        private bool _secondLaneEnabled = true;

        private bool _thirdLaneEnabled = true;
        private bool _fourthLaneEnabled = true;
```

File: Assets/Scripts/Framework/BeatMapping/TimeStamper.cs [hunk 6]
```diff
@@ -280,6 +318,7 @@ public void TryExportTimeStamps(string overrideSongTitle = "")
 
             obj.songDebugLineData = exportedLineData;
             obj.timeStamps = sortedExportedTimeStamps.ToArray();
+            obj.laneIDs = exportedLaneIDs.ToArray(); 
 
             //Get the song title for the asset name.
             string assetName = overrideSongTitle == string.Empty ? _waveformDrawer.CurrentSongTitle : overrideSongTitle;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStamper.cs]:
```suggestion
            string assetName = overrideSongTitle == string.Empty 
                      ? _waveformDrawer.CurrentSongTitle 
                      : overrideSongTitle;
```

**Review by Joost-Robat — CHANGES_REQUESTED** [review]
Some changes are required, but it's looking really good.

File: Assets/Art/2D/UI/GameScreen/S_HPBalk.png
Comment (lucasvdiepen) [inline Assets/Art/2D/UI/GameScreen/S_HPBalk.png:1]: Change to `HP Bar`

File: Assets/Art/2D/UI/LevelPlay/S_BalkForStars.png
Comment (lucasvdiepen) [inline Assets/Art/2D/UI/LevelPlay/S_BalkForStars.png:1]: Change to `BarForStars`

File: Assets/Art/2D/UI/LevelPlay/S_MediumStarOranje.png
Comment (lucasvdiepen) [inline Assets/Art/2D/UI/LevelPlay/S_MediumStarOranje.png:1]: Change to `S_MediumStarOrange`

File: Assets/Editor/ScriptableObjects/LevelData/SongCovers/BadTimePiggies.png
Comment (lucasvdiepen) [inline Assets/Editor/ScriptableObjects/LevelData/SongCovers/BadTimePiggies.png:1]: Add `S_` prefix

File: Assets/Editor/ScriptableObjects/LevelData/SongCovers/BullseyeByKDrew.jpg
Comment (lucasvdiepen) [inline Assets/Editor/ScriptableObjects/LevelData/SongCovers/BullseyeByKDrew.jpg:1]: Add `S_` prefix

File: Assets/Editor/ScriptableObjects/LevelData/SongCovers/FeelMyHeartByChestbrah.jpg
Comment (lucasvdiepen) [inline Assets/Editor/ScriptableObjects/LevelData/SongCovers/FeelMyHeartByChestbrah.jpg:1]: Add `S_` prefix

File: Assets/Scripts/Framework/Buttons/PlayLevel/PlayLevelButton.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+using ProefExamen.Framework.StateMachine.States;
+
+namespace ProefExamen.Framework.Buttons.PlayLevel
+{
+    public class PlayLevelButton : BasicButton
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Buttons/PlayLevel/PlayLevelButton.cs:8]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/GameState.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+using System.Collections;
+
+using ProefExamen.Framework.Gameplay.LaneSystem;
+
+namespace ProefExamen.Framework.StateMachine.States
+{
+    public class GameState : MenuState
+    {
+        public override IEnumerator OnStateEnter()
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/GameState.cs]: Both missing summaries

File: Assets/Scripts/Framework/StateMachine/States/AdState.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+using ProefExamen.Framework.StateMachine.Attributes;
+
+namespace ProefExamen.Framework.StateMachine.States
+{
+    [ParentState(typeof(GameState))]
+    public class AdState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/AdState.cs:9]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/LevelSelectorState.cs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+namespace ProefExamen.Framework.StateMachine.States
+{
+    public class LevelSelectorState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/LevelSelectorState.cs:7]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/LoseState.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+using ProefExamen.Framework.StateMachine.Attributes;
+
+namespace ProefExamen.Framework.StateMachine.States
+{
+    [ParentState(typeof(GameState))]
+    public class LoseState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/LoseState.cs:9]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/PauseState.cs [hunk 12]
```diff
@@ -0,0 +1,10 @@
+using ProefExamen.Framework.StateMachine.Attributes;
+
+namespace ProefExamen.Framework.StateMachine.States
+{
+    [ParentState(typeof(GameState))]
+    public class PauseState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/PauseState.cs:9]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/PlayLevelState.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+namespace ProefExamen.Framework.StateMachine.States
+{
+    public class PlayLevelState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/PlayLevelState.cs:6]: Missing summary

File: Assets/Scripts/Framework/StateMachine/States/WinState.cs [hunk 14]
```diff
@@ -0,0 +1,10 @@
+using ProefExamen.Framework.StateMachine.Attributes;
+
+namespace ProefExamen.Framework.StateMachine.States
+{
+    [ParentState(typeof(GameState))]
+    public class WinState : MenuState
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/StateMachine/States/WinState.cs:9]: Missing summary

File: Assets/Scripts/Framework/UI/Levels/LevelInfoUpdater.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
+using UnityEngine;
+using TMPro;
+using Image = UnityEngine.UI.Image;
+
+using ProefExamen.Framework.Gameplay.Level;
+using ProefExamen.Framework.Gameplay.Values;
+using ProefExamen.Framework.UI;
+
+public class LevelInfoUpdater : MonoBehaviour
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/UI/Levels/LevelInfoUpdater.cs]: Missing namespace and summary

File: Assets/Scripts/Framework/UI/Levels/LevelSelectorSpawner.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+using UnityEngine;
+
+using ProefExamen.Framework.Gameplay.Level;
+using ProefExamen.Framework.Gameplay.Values;
+using ProefExamen.Framework.Buttons.LevelSelector;
+using ProefExamen.Framework.UI.TextUpdaters;
+
+namespace ProefExamen.Framework.UI.Level
+{
+    public class LevelSelectorSpawner : MonoBehaviour
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/UI/Levels/LevelSelectorSpawner.cs:14]: Missing summary

File: Assets/Scripts/Framework/UI/MenuStateUpdater.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+namespace ProefExamen.Framework.UI
+{
+    public class MenuStateUpdater : AbstractSingleton<MenuStateUpdater>
+    {
+        public Action<Difficulty> OnDifficultyChanged;
+
+        public Action<int> OnHighScoreChanged;
+
+        public Action<int> OnSelectedLevelIDChanged;
+    }
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/UI/MenuStateUpdater.cs]: Missing summaries

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Framework/StateMachine/States/GameState.cs (hunk 8 above)
Comment (lucasvdiepen) [reply Assets/Scripts/Framework/StateMachine/States/GameState.cs]: `OnStateEnter()` is still missing a summary

File: Assets/Scripts/Framework/UI/Levels/LevelInfoUpdater.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
+using Image = UnityEngine.UI.Image;
+using UnityEngine;
+using TMPro;
+
+using ProefExamen.Framework.Gameplay.Level;
+using ProefExamen.Framework.Gameplay.Values;
+
+namespace ProefExamen.Framework.UI.Level
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/UI/Levels/LevelInfoUpdater.cs]:
```suggestion
namespace ProefExamen.Framework.UI.Levels
```

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

**Review by lucasvdiepen — APPROVED** [review]

**Review by Joost-Robat — COMMENTED** [review]
Only a ternary statement that isn't according to code conventions but the rest looks good.

File: Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs [hunk 19]
```diff
@@ -53,22 +85,40 @@ private void OnDrawGizmos()
...
+                if(lane1Disabled || lane2Disabled || lane3Disabled || lane4Disabled)
+                    continue;   
 
                 Vector2 offset = new(_gizmoSpacing, 0);
-                Gizmos.color = _timeStamper.TimeStamps[i].isSelected ? _selectedTimeStampColor : _timeStampColor;
+
+                bool isSelected = _timeStamper.TimeStamps[i].isSelected;
+                Color normalGizmoColor = _timeStampLaneColor[_timeStamper.TimeStamps[i].laneID];
+
+                Gizmos.color = isSelected ? _selectedTimeStampColor : normalGizmoColor;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/BeatMapping/TimeStampDebugger.cs]:
```suggestion
                Gizmos.color = isSelected 
                        ? _selectedTimeStampColor 
                        : normalGizmoColor;
```

**Review by Joost-Robat — APPROVED** [review]
Looks good, ready for merge.

**Pull request merged**